### PR TITLE
Use application name as default Kafka consumer group.id

### LINKED ISSUE #15632: Use application name (if set) as Kafka group.id
**Description**

At the moment, the group.id is:

* not set in prod mode - meaning that the user must set it
* is generated to be a unique id in dev mode - to facilitate the hot reload

It would be great to set the group.id to the application name if that name is set. 

**Implementation ideas**

If the application name is set, add the "kafka.group.id" property.

### PR DESCRIPTION
**Pull request by Ladicek**
Resolves #15632

### CODE REVIEW

**Review by cescoffier — APPROVED** [review]
Looks good to me! 

I just made a small comment.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java [hunk 1]
```diff
@@ -13,32 +13,40 @@
...
                 .sorted()
                 .forEach(name -> {
-                    final String key = name.substring(configPrefix.length() + 1).toLowerCase().replaceAll("[^a-z0-9.]", ".");
+                    final String key = name.substring(CONFIG_PREFIX.length() + 1).toLowerCase().replaceAll("[^a-z0-9.]", ".");
                     final String value = config.getOptionalValue(name, String.class).orElse("");
                     properties.put(key, value);
                 });
 
+        if (!properties.containsKey(GROUP_ID) && app.name.isPresent()) {
+            properties.put(GROUP_ID, app.name.get());
```
Comment (cescoffier) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java:47]: Should we "kafka.group.id" not just "group.id"

**Review by cescoffier — CHANGES_REQUESTED** [review]

**Comment by Ladicek** [comment]
Draft to let CI run in my fork.

I was thinking how to add a test for this, but couldn't find anything. I don't think SmallRye Reactive Messaging lets me access the internals, through which I could verify if the group ID is really set. I did verify it manually with a simple example application though.

Also, #15632 says that a generated ID is only used in dev mode -- I couldn't find that anywhere in Quarkus code, and the SmallRye Reactive Messaging code generated a UUID unconditionally. So I think #15632 is just mistaken.

File: docs/src/main/asciidoc/kafka.adoc [hunk 2]
```diff
@@ -915,7 +915,11 @@ Type: _string_ | true |
 
 Type: _int_ | false | `1`
 
-| *group.id* | A unique string that identifies the consumer group the application belongs to. If not set, a unique, generated id is used
+| *group.id* | A unique string that identifies the consumer group the application belongs to.
+
+If not set, defaults to the application name as set by the `quarkus.application.name` configuration property.
+
+If that is not set either, a unique, generated id is used.
```
[the remark this answers is not in the document]
Comment (Ladicek) [reply docs/src/main/asciidoc/kafka.adoc:922]: Makes sense, I'll add something like that.

**Comment by geoand** [comment]
Interesting coincidence, I just opened https://github.com/quarkusio/quarkus/pull/15680 :)

File: docs/src/main/asciidoc/kafka.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ladicek) [reply docs/src/main/asciidoc/kafka.adoc:922]: Done.

**Comment by Ladicek** [comment]
@geoand I'm fine with rebasing this on top of #15680 as well, so if you get to merging your first, please do :-)

**Comment by geoand** [comment]
No need, I'll wait :)

**Comment by Ladicek** [comment]
CI passed in my fork.

**Comment by wernert75** [comment]
In principle, I think that the change is good. 
However, it now contradicts the doc: 

"A unique string that identifies the consumer group the application belongs to. If not set, a unique, generated id is used
Type: string"

In a special case I need the previous behavior of a randomly generated group.id. How can I restore this behavior?

**Comment by Ladicek** [comment]
It shouldn't contradict the doc, as I updated the doc in the same commit :-)

If you want your group ID random, you can manually generate a random string and put it into an env var (`KAFKA_GROUP_ID`), for example.

**Comment by cescoffier** [comment]
As I said on zulip, using a generated UUID needs to be a conscientious decision.

**Comment by wernert75** [comment]
Thank you for your answer. 

I agree with you. I think it's a good change. And using a generated UUID schould be a conscientious decision.

But the doc seems to be not updated on the website
https://quarkus.io/guides/kafka 

And a mention in the Migration Guide would have saved me a lot of time. ;)  

Thanks again. I will solve it via the configuration. 

Perhaps you will also consider again whether an explicit configuration variable that activates the explicit use of a UUID is possible.

Basically, the issue is that several replicas of one and the same microservice get different Group.IDs, to broadcast a message.

It would be ideal if there could be a kind of "pool" for Group.ids. For example, MyApplication1 ... MyApplication5. If someone has an idea how to solve the problem of assigning exactly one of these GroupIds to each replica via a configuration, I would be very happy to hear about it. 

Translated with www.DeepL.com/Translator (free version)

**Comment by Ladicek** [comment]
Hmm I wonder how that website is updated, maybe I updated a wrong file? Humm...

**Comment by Ladicek** [comment]
Ah ah, now I know who to blame: https://github.com/quarkusio/quarkus/commit/8a56a30b516f72ac7ed7e177ac118a831ef6d46a

**Comment by Ladicek** [comment]
I'll submit a PR to re-add the new docs.

**Comment by Ladicek** [comment]
Submitted #16262.
